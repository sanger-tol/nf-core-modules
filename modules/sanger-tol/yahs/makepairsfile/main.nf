process YAHS_MAKEPAIRSFILE {
    tag "${meta.id}"
    label "process_medium"

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/dd/ddcb4312a2df1f479502951d398f1bb983f82728446784f8fa9c4e91074ffdb7/data' :
        'community.wave.seqera.io/library/samtools_yahs_coreutils_gawk:4f936f9704189766' }"

    input:
    tuple val(meta), path(scaffolds_fai), path(scaffolds_agp), path(contigs_fai), path(contigs_contacts)

    output:
    tuple val(meta), path("*.pairs.gz"), emit: pairs
    path "versions.yml"                , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def buffer_mem  = (task.memory.toGiga() / 2).round()
    """
    gawk '
        BEGIN {
            OFS = "\t"
            print "## pairs format v1.0"
        }
        { print "#chromsize:", \$1, \$2 }
        END {
            print "#columns:", "readID", "chr1", "pos1", "chr2", "pos2", "strand1", "strand2"
        }' ${scaffolds_fai} > ${prefix}.pairs

    juicer pre \\
        ${contigs_contacts} \\
        ${scaffolds_agp} \\
        ${contigs_fai} |\\
    gawk '
        BEGIN { OFS = "\t" }
        \$3 > 0 && \$7 > 0 {print ".", \$2, \$3, \$6, \$7, ".", "."}
    ' |\\
    LC_ALL=C sort -k2,2d -k4,4d \\
        -S${buffer_mem}G \\
        -T . \\
        --parallel=${task.cpus} >>\\
        ${prefix}.pairs

    bgzip -@${task.cpus} ${prefix}.pairs

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        juicer_pre: \$(juicer_pre --version)
        gawk: \$(awk -Wversion | sed '1!d; s/.*Awk //; s/,.*//')
        sort: \$(sort --version | sed '1!d' | grep -o -E "[0-9]+(\\.[0-9]+)+")
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}.pairs.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        juicer_pre: \$(juicer_pre --version)
        gawk: \$(awk -Wversion | sed '1!d; s/.*Awk //; s/,.*//')
        sort: \$(sort --version | sed '1!d' | grep -o -E "[0-9]+(\\.[0-9]+)+")
    END_VERSIONS
    """
}
