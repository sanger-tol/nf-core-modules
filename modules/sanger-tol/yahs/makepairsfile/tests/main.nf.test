nextflow_process {

    name "Test Process YAHS_MAKEPAIRSFILE"
    script "../main.nf"
    process "YAHS_MAKEPAIRSFILE"

    tag "modules"
    tag "modules_sangertol"
    tag "yahs"
    tag "yahs/makepairsfile"

    test("Undibacterium unclassified - YaHS bin") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/scaffolding/baUndUnlc1_hic_phased_hap2_scaffolds_final.fa.fai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'baUndUnlc1_hic_phased_hap2_scaffolds_final.agp', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'baUndUnlc1.hic.hap2.p_ctg.fa.fai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'baUndUnlc1_hic_phased_hap2.bin', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success               },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("Undibacterium unclassified - YaHS bin - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/scaffolding/baUndUnlc1_hic_phased_hap2_scaffolds_final.fa.fai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'baUndUnlc1_hic_phased_hap2_scaffolds_final.agp', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'baUndUnlc1.hic.hap2.p_ctg.fa.fai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'baUndUnlc1_hic_phased_hap2.bin', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success               },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
