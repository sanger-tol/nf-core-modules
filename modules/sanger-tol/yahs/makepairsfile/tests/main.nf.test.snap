{
    "Undibacterium unclassified - YaHS bin": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pairs.gz:md5,c1be641ec6fb8cfd1b0fb674a12b51e5"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a365bc19a25174bdf6a227418db2ac03"
                ],
                "pairs": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pairs.gz:md5,c1be641ec6fb8cfd1b0fb674a12b51e5"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a365bc19a25174bdf6a227418db2ac03"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-28T16:06:08.417556224"
    },
    "Undibacterium unclassified - YaHS bin - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pairs.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a365bc19a25174bdf6a227418db2ac03"
                ],
                "pairs": [
                    [
                        {
                            "id": "test"
                        },
                        "test.pairs.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a365bc19a25174bdf6a227418db2ac03"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-28T16:06:16.432728937"
    }
}